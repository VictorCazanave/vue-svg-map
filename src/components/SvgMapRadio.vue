<script setup lang="ts">
import type { SVGAttributes } from 'vue'
import type { Map, Location } from '../types'
import { onMounted, useTemplateRef } from 'vue'
import SvgMap from './SvgMap.vue'

defineProps<{map: Map}>()

// Id of selected location
const model = defineModel<string>({ default: () => null })

const svgComponent = useTemplateRef('svg')
let locationElts: SVGPathElement[]

onMounted(() => {
  if(svgComponent.value) {
    // Store all location elements to handle focus
    locationElts = svgComponent.value.$el.querySelectorAll('path')
  }
})

/**
 * Indicate whether location is selected
 *
 * @param location - Location to check
 * @returns True if location is selected
 */
function isLocationSelected(location: Location): boolean {
  return model.value === location.id
}

/**
 * Get tabindex value of location
 *
 * @param location - Location of which to get tabindex
 * @param index - Index of location
 * @returns Value of tabindex HTML attribute
 */
function getLocationTabindex(location: Location, index: number): string {
  if (model.value) {
    // Only selected location is focusable
    return isLocationSelected(location) ? '0' : '-1'
  }

  // Only first location is focusable by default
  return index === 0 ? '0' : '-1'
}

/**
 * Select a location
 *
 * @param location - DOM node of location to select
 */
function selectLocation(locationElt: SVGPathElement): void {
  // Select only if location is not already selected
  if (locationElt.id !== model.value) {
    model.value = locationElt.id
    locationElt.focus()
  }
}

/**
 * Select next or first location
 *
 * @param event - Triggered event
 */
function selectNextLocation(event: KeyboardEvent): void {
  const target = event.target as SVGElement

  if(target) {
    // Next sibling can be anything in `after` slot
    const nextElementSibling = target.nextElementSibling
    const nextLocationElt = nextElementSibling?.nodeName === 'path'
      ? nextElementSibling as SVGPathElement
      : locationElts[0]

    selectLocation(nextLocationElt)
  }
}

/**
 * Select previous or last location
 *
 * @param event - Triggered event
 */
function selectPreviousLocation(event: KeyboardEvent): void {
  const target= event.target as SVGElement

  if(target) {
    // Previous sibling can be anything in `before` slot
    const previousElementSibling = target.previousElementSibling
    const previousLocation = previousElementSibling?.nodeName === 'path'
      ? previousElementSibling as SVGPathElement :
      locationElts[locationElts.length - 1]

      selectLocation(previousLocation)
  }
}

/**
 * Get HTML attributes of location depending on its state
 *
 * @return HTML attributes of location
 */
function getLocationAttributes(location: Location, index: number): SVGAttributes {
  return {
    role: 'radio',
    tabindex: getLocationTabindex(location, index),
    'aria-checked': isLocationSelected(location),
  }
}
</script>

<template>
  <!-- NOTE: Use .stop modifier to prevent event bubbling and calling function twice -->
	<SvgMap
    ref="svg"
		v-bind="$attrs"
    :map="map"
		role="radiogroup"
    :locationAttributes="getLocationAttributes"
		@click.stop="selectLocation($event.target)"
		@keydown.space.prevent.stop="selectLocation($event.target)"
		@keydown.down.prevent.stop="selectNextLocation"
		@keydown.right.prevent.stop="selectNextLocation"
		@keydown.up.prevent.stop="selectPreviousLocation"
		@keydown.left.prevent.stop="selectPreviousLocation"
	>
		<!-- Pass down slots to SvgMap: https://stackoverflow.com/a/50892881/9826498 -->
		<template
			v-for="(_, slotName) of $slots"
			v-slot:[slotName]="scope"
		>
			<slot
				v-bind="scope"
				:name="slotName"
			/>
		</template>
	</SvgMap>
</template>
