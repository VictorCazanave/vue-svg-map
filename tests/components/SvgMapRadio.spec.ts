import { mount,  VueWrapper, DOMWrapper } from '@vue/test-utils'
import map from './mocks/map'
import SvgMapRadio from '../../src/components/SvgMapRadio.vue'

describe('SvgMapRadio component', () => {
	const modelValue = 'id1'
  let wrapper: VueWrapper
  let locationElt0: DOMWrapper<SVGPathElement>
	let locationElt1: DOMWrapper<SVGPathElement>
	let locationElt2: DOMWrapper<SVGPathElement>

	beforeEach(() => {
		wrapper = mount(SvgMapRadio, { props: { map, modelValue } })

    const locationElts: DOMWrapper<SVGPathElement>[] = wrapper.findAll('path')
    locationElt0 = locationElts.at(0)!
    locationElt1 = locationElts.at(1)!
    locationElt2 = locationElts.at(2)!
	})

	it('renders svg map as a group of radio buttons', () => {
		expect(wrapper.element).toMatchSnapshot()
	})

	it('selects location of inital model value', () => {
		expect(locationElt0.attributes('aria-checked')).toBe('false')
		expect(locationElt1.attributes('aria-checked')).toBe('true')
		expect(locationElt2.attributes('aria-checked')).toBe('false')
	})

	it('selects location of new model value', async () => {
		await wrapper.setProps({ modelValue: 'id0' })

		expect(locationElt0.attributes('aria-checked')).toBe('true')
		expect(locationElt1.attributes('aria-checked')).toBe('false')
		expect(locationElt2.attributes('aria-checked')).toBe('false')
	})

	it('sets only selected location as focusable', () => {
		expect(locationElt0.attributes('tabindex')).toBe('-1')
		expect(locationElt1.attributes('tabindex')).toBe('0')
		expect(locationElt2.attributes('tabindex')).toBe('-1')
	})

	it('sets only first location as focusable when no selected location', async () => {
		await wrapper.setProps({ modelValue: null })

		expect(locationElt0.attributes('tabindex')).toBe('0')
		expect(locationElt1.attributes('tabindex')).toBe('-1')
		expect(locationElt2.attributes('tabindex')).toBe('-1')
	})

	it('emits update:modelValue event with new selected location when clicking on an unselected location', async () => {
		await locationElt2.trigger('click')

		expect(wrapper.emitted('update:modelValue')).toEqual([['id2']])
	})

	it('does not emit update:modelValue event when clicking on selected location', async () => {
		await locationElt1.trigger('click')

		expect(wrapper.emitted('update:modelValue')).toBeUndefined()
	})

	it('emits update:modelValue event with new selected location when pressing spacebar on an unselected location', async () => {
		await locationElt2.trigger('keydown.space')

		expect(wrapper.emitted('update:modelValue')).toEqual([['id2']])
	})

	it('does not emit update:modelValue event when pressing spacebar on selected location', async () => {
		await locationElt1.trigger('keydown.space')

		expect(wrapper.emitted('update:modelValue')).toBeUndefined()
	})

	it('emits update:modelValue event with next location when pressing down arrow on selected location', async () => {
		await locationElt1.trigger('keydown.down')

		expect(wrapper.emitted('update:modelValue')).toEqual([['id2']])
	})

	it('emits update:modelValue event with first location when pressing down arrow on selected last location', async () => {
		await wrapper.setProps({ value: 'id2' })
		await locationElt2.trigger('keydown.down')

		expect(wrapper.emitted('update:modelValue')).toEqual([['id0']])
	})

	it('emits update:modelValue event with next location when pressing right arrow on selected location', async () => {
		await locationElt1.trigger('keydown.right')

		expect(wrapper.emitted('update:modelValue')).toEqual([['id2']])
	})

	it('emits update:modelValue event with first location when pressing right arrow on selected last location', async () => {
		await wrapper.setProps({ value: 'id2' })
		locationElt2.trigger('keydown.right')

		expect(wrapper.emitted('update:modelValue')).toEqual([['id0']])
	})

	it('emits update:modelValue event with previous location when pressing up arrow on selected location', () => {
		locationElt1.trigger('keydown.up')

		expect(wrapper.emitted('update:modelValue')).toEqual([['id0']])
	})

	it('emits update:modelValue event with last location when pressing up arrow on selected first location', async () => {
		await wrapper.setProps({ value: 'id0' })
		locationElt0.trigger('keydown.up')

		expect(wrapper.emitted('update:modelValue')).toEqual([['id2']])
	})

	it('emits update:modelValue event with previous location when pressing left arrow on selected location', () => {
		locationElt1.trigger('keydown.left')

		expect(wrapper.emitted('update:modelValue')).toEqual([['id0']])
	})

	it('emits update:modelValue event with last location when pressing left arrow on selected first location', async () => {
		await wrapper.setProps({ value: 'id0' })
		locationElt0.trigger('keydown.left')

		expect(wrapper.emitted('update:modelValue')).toEqual([['id2']])
	})
})
