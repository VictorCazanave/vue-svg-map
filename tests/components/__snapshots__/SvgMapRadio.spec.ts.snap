// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SvgMapRadio component > renders svg map as a group of radio buttons 1`] = `
<div
  data-v-app=""
>
  
  <!-- NOTE: Use .stop modifier to prevent event bubbling and calling function twice -->
  <svg
    aria-label="label"
    class="svg-map"
    role="radiogroup"
    viewBox="viewBox"
    xmlns="http://www.w3.org/2000/svg"
  >
    
    
    
    <path
      aria-checked="false"
      aria-label="name0"
      class="svg-map__location"
      d="path0"
      id="id0"
      role="radio"
      tabindex="-1"
    />
    <path
      aria-checked="true"
      aria-label="name1"
      class="svg-map__location"
      d="path1"
      id="id1"
      role="radio"
      tabindex="0"
    />
    <path
      aria-checked="false"
      aria-label="name2"
      class="svg-map__location"
      d="path2"
      id="id2"
      role="radio"
      tabindex="-1"
    />
    
    
    
  </svg>
  
</div>
`;
